plutus-benchmark/src/Plutus/Benchmark/Queens.hs

```diff
@@ -0,0 +1,553 @@
+   five different algorithms, and the original behaviour was to try each of
+   these on a board if size NxN, where N is a parameter supplied by the user.
+   The Haskell version can run all five algorithms one after the other hundreds
+   of times on an 8x8 board in a few seconds.
+
+   Compiling into Plutus, I tried to run all five algorithms one after the other
+   (in a sinlge run of the program) on a 5x5 board on the CEK machine on an 8 GB
+   machine.  The program was just able to complete in about 23s, but with some
+   swapping.
+
+   The resource consumption isn't so bad for the individual algorithms.  Here
+   are figures for the time and memory usage for one run of the Plutus version
+   of each algorithm on a 5x5 board on the same machine:
+
+   bt:    2.9s, 1.2 GB
+   bm:    2.5s, 830 MB
+   bjbt:  2.7s, 933 MB
+   bjbt': 3.0s, 1.1 GB
+   fc:    7.5s, 3.2 GB
+
+   I suggest keeping the entire program, but only using a single algorithm for
+   benchmarking for the time being.  If we're investigate the behaviour of
+   Plutus Core evaluators at some point it might be useful to have the whole
+   range of algorithms available for experiments.
+-}
+
+
+-- The different algorithms implemented in this file. The program iterates the
+-- solver over this list of algortihms.
+
+{-# INLINABLE algorithms #-}
+algorithms :: [Labeler]
+-- algorithms = [bt, bm, bjbt, bjbt', fc]
+algorithms = [bm]
+
+data Algorithm = Bt
+               | Bm
+               | Bjbt
+               | Bjbt'
+               | Fc
+               deriving (Show)
+
+{-# INLINABLE lookupAlgorithm #-}
+lookupAlgorithm :: Algorithm -> Labeler
+lookupAlgorithm Bt    = bt
+lookupAlgorithm Bm    = bm
+lookupAlgorithm Bjbt  = bjbt
+lookupAlgorithm Bjbt' = bjbt'
+lookupAlgorithm Fc    = fc
+
+-- The main input parameter used by the Plutus version, the size of the board (n).
+boardSize :: Integer
+boardSize = 5
+
+{-# INLINABLE nqueens #-}
+nqueens :: Integer -> Labeler -> Integer
+nqueens n algorithm = length (search algorithm (queens n))
+
+mkQueensTerm :: Integer -> [Algorithm] -> Term Name DefaultUni ()
+mkQueensTerm sz algs =
+  let (Program _ _ code) =
+        Tx.getPlc $ $$(Tx.compile [||
+          \sz algs -> map (nqueens sz)
+                          (map lookupAlgorithm algs) ||])
+          `Tx.applyCode` Tx.liftCode sz
+          `Tx.applyCode` Tx.liftCode algs
+  in code
+
+main :: IO ()
+main = main1
+
+main1 :: IO ()  -- Compile a Plutus Core version
+main1 = do
+  let code = Tx.getPlc $ $$(Tx.compile [|| map (nqueens boardSize) algorithms  ||])
+  mapM_ putStrLn $ unindent . PLC.prettyPlcClassicDebug $ code
+
+main2 :: IO()  -- Haskell version
+main2 = do
+  args <- getArgs
+  case args of
+    [] -> putStrLn "Integer parameter expected"
+    arg:_ -> do
+              let n = read arg :: Integer
+                  try algorithm = print (nqueens n algorithm)
+              forM_ [1..240] $ const $ do
+                sequence_ (map try algorithms)
```

Comment:
Nitpicking: `sequence_ . map f` is better known as `traverse_ f`.

Reply:
Yeah, but that's from 1999.